File: Kern/Tests/DuoKernTests/BoxGrowthTests.swift
import Foundation
import Testing
import DuoKern

@Suite("Box growth: budget, health gate, introduction")
struct BoxGrowthTests {
    let cal = Box.calendar
    let now = Box.day1

    @Test("day-one bootstrap: empty box fills the learning pool despite 0 active cards")
    func dayOneBootstrap() {
        let state = Box.state(cards: (1...10).map { Box.word($0) }, Box.config(maxLearning: 5))
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan.reviews.isEmpty)
        #expect(plan.unlockedPhrases.isEmpty)
        #expect(plan.newWords == ["w01", "w02", "w03", "w04", "w05"])
        #expect(!plan.isEmpty)
    }

    @Test("growth blocked when relearning share >= 20% with >= 10 active")
    func relearningGateBlocks() {
        var state = Box.state(cards: (1...12).map { Box.word($0) })
        let future = now.addingTimeInterval(5 * 86_400)
        for n in 1...8 {
            Box.inject(&state, Box.sched(String(format: "w%02d", n), due: future,
                                         lastReview: now.addingTimeInterval(-86_400)))
        }
        for n in 9...10 {
            Box.inject(&state, Box.sched(String(format: "w%02d", n),
                                         phase: .relearning, due: future,
                                         lastReview: now.addingTimeInterval(-86_400)))
        }
        // 2 of 10 active relearning = 20% → gate closed
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan.newWords.isEmpty)
        #expect(plan.unlockedPhrases.isEmpty)
        let stats = BoxEngine.statistics(state: state, now: now, calendar: cal)
        #expect(stats.newSlotsAvailable == 0)

        // Drop to 1 of 10 (10%) → gate open again
        var healthy = state
        Box.inject(&healthy, Box.sched("w10", phase: .review, due: future,
                                       lastReview: now.addingTimeInterval(-86_400)))
        let plan2 = BoxEngine.composeSession(state: healthy, now: now, calendar: cal)
        #expect(plan2.newWords == ["w11", "w12"])
    }

    @Test("relearning sub-gate passes below 10 active cards (bootstrap clause)")
    func relearningGateSkippedWhenSmall() {
        var state = Box.state(cards: (1...8).map { Box.word($0) })
        let future = now.addingTimeInterval(5 * 86_400)
        for n in 1...2 {
            Box.inject(&state, Box.sched("w0\(n)", phase: .relearning, due: future,
                                         lastReview: now.addingTimeInterval(-86_400)))
        }
        for n in 3...5 {
            Box.inject(&state, Box.sched("w0\(n)", due: future, lastReview: now.addingTimeInterval(-86_400)))
        }
        // 2 of 5 relearning = 40%, but activeCount < 10 → still introduces
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan.newWords == ["w06", "w07", "w08"])
    }

    @Test("growth blocked when projected post-session backlog >= dueSoftCap")
    func backlogGateBlocks() {
        var state = Box.state(cards: (1...70).map { Box.word($0) })
        for n in 1...61 {
            let id = String(format: "w%02d", n)
            Box.inject(&state, Box.sched(id, due: now.addingTimeInterval(-Double(n) * 60),
                                         lastReview: now.addingTimeInterval(-86_400)))
        }
        // 61 due − 30 sessionCap = 31 ≥ dueSoftCap 30 → closed; no slots reserved
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan.newWords.isEmpty)
        #expect(plan.reviews.count == 30)
    }

    @Test("pool budget tracks the learning load: compose, answer some, recompose")
    func budgetAcrossRecomposition() {
        var state = Box.state(cards: (1...10).map { Box.word($0) }, Box.config(maxLearning: 5))
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan.newWords.count == 5)

        // Answering .good moves each card into .learning, filling the pool.
        for id in plan.newWords.prefix(3) {
            state = BoxEngine.answer(state: state, cardID: id, rating: .good, now: now, calendar: cal)
        }
        let plan2 = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan2.newWords == ["w04", "w05"]) // 5 pool − 3 learning = 2 free

        for id in plan2.newWords {
            state = BoxEngine.answer(state: state, cardID: id, rating: .good, now: now, calendar: cal)
        }
        let plan3 = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan3.newWords.isEmpty) // pool full (5 in .learning)
    }

    @Test("pool refills as learning cards graduate to review")
    func poolRefillsOnGraduation() {
        var state = Box.state(cards: (1...10).map { Box.word($0) }, Box.config(maxLearning: 3))
        // Fill the pool: 3 cards into .learning.
        for id in ["w01", "w02", "w03"] {
            state = BoxEngine.answer(state: state, cardID: id, rating: .good, now: now, calendar: cal)
        }
        #expect(BoxEngine.composeSession(state: state, now: now, calendar: cal).newWords.isEmpty)

        // Graduate w01 (answer its learning step .good → .review) frees one slot.
        let step = now.addingTimeInterval(700)
        state = BoxEngine.answer(state: state, cardID: "w01", rating: .good, now: step, calendar: cal)
        let refilled = BoxEngine.composeSession(state: state, now: step, calendar: cal)
        #expect(refilled.newWords == ["w04"]) // one graduated → one new pulled in
    }

    @Test("introduction at first answer: composing without answering burns no budget")
    func compositionBurnsNoBudget() {
        let state = Box.state(cards: (1...10).map { Box.word($0) })
        var current = state
        for _ in 0..<5 {
            _ = BoxEngine.composeSession(state: current, now: now, calendar: cal)
        }
        #expect(current == state)
        #expect(current.newIntroduced.isEmpty)
        #expect(current.scheduling.isEmpty)
    }

    @Test("pool budget re-checked defensively at answer time; a full pool blocks introduction")
    func budgetDefensiveAtAnswer() {
        var state = Box.state(cards: (1...5).map { Box.word($0) }, Box.config(maxLearning: 2))
        state = BoxEngine.answer(state: state, cardID: "w01", rating: .good, now: now, calendar: cal)
        state = BoxEngine.answer(state: state, cardID: "w02", rating: .good, now: now, calendar: cal)
        // Pool full (2 in .learning) → introducing a third is a no-op.
        let blocked = BoxEngine.answer(state: state, cardID: "w03", rating: .good, now: now, calendar: cal)
        #expect(blocked == state)
        #expect(blocked.scheduling.count == 2)

        // Graduate w01 → a slot frees → the same answer now succeeds.
        let step = now.addingTimeInterval(700)
        var freed = BoxEngine.answer(state: state, cardID: "w01", rating: .good, now: step, calendar: cal)
        freed = BoxEngine.answer(state: freed, cardID: "w03", rating: .good, now: step, calendar: cal)
        #expect(freed.scheduling.count == 3)
    }

    @Test("enqueued ids lead within the pool budget; phrase enqueue pulls missing components first")
    func enqueuePriority() {
        var state = Box.state(cards: (1...10).map { Box.word($0) }, Box.config(maxLearning: 5))
        state = BoxEngine.enqueue(state: state, cardIDs: ["w07"])
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        // w07 leads but shares the pool budget (5): 1 enqueued + 4 automatic = 5.
        #expect(plan.newWords == ["w07", "w01", "w02", "w03", "w04"])

        var withPhrase = Box.state(cards: (1...6).map { Box.word($0) }
            + [Box.phrase("p1", components: ["w05", "w06"])], Box.config(maxLearning: 5))
        withPhrase = BoxEngine.enqueue(state: withPhrase, cardIDs: ["p1"])
        #expect(withPhrase.enqueued == ["w05", "w06", "p1"])
        let plan2 = BoxEngine.composeSession(state: withPhrase, now: now, calendar: cal)
        // locked phrase never enters, even enqueued; its components lead, then
        // automatic fills the rest of the pool budget (2 enqueued + 3 automatic).
        #expect(plan2.newWords == ["w05", "w06", "w01", "w02", "w03"])
    }

    @Test("enqueued cards respect the load throttle: a full pool defers a pack, it is not dumped")
    func enqueuedRespectsLoadThrottle() {
        var state = Box.state(cards: (1...10).map { Box.word($0) }, Box.config(maxLearning: 2))
        // Pack three cards, but the pool budget is only 2 → at most 2 surface.
        state = BoxEngine.enqueue(state: state, cardIDs: ["w06", "w07", "w08"])
        let plan = BoxEngine.composeSession(state: state, now: now, calendar: cal)
        #expect(plan.newWords == ["w06", "w07"]) // drips in at pool rate, not all three

        // Fill the pool by answering them → next session defers the rest.
        state = BoxEngine.answer(state: state, cardID: "w06", rating: .good, now: now, calendar: cal)
        state = BoxEngine.answer(state: state, cardID: "w07", rating: .good, now: now, calendar: cal)
        #expect(BoxEngine.composeSession(state: state, now: now, calendar: cal).newWords.isEmpty)
        #expect(state.enqueued == ["w08"]) // still waiting for a slot
    }

    @Test("answering an enqueued card removes it from the queue")
    func answerDequeues() {
        var state = Box.state(cards: (1...3).map { Box.word($0) })
        state = BoxEngine.enqueue(state: state, cardIDs: ["w02"])
        state = BoxEngine.answer(state: state, cardID: "w02", rating: .good, now: now, calendar: cal)
        #expect(state.enqueued.isEmpty)
    }
}

File: kmp/kern/src/commonTest/kotlin/net/spross/kern/box/BoxGrowthTests.kt
package net.spross.kern.box

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import net.spross.kern.model.CardPhase
import net.spross.kern.model.Rating

/** Growth: pool budget, health gate, enqueue — everything denominated in cards. */
class BoxGrowthTests {
    private val now = Box.day1

    @Test
    fun dayOneBootstrapFillsThePool() {
        val state = Box.state((1..10).map { Box.word(it) }, Box.config(maxLearning = 5))
        val plan = Box.candidates(state)
        assertTrue(plan.unlockedPhrases.isEmpty())
        assertEquals((1..5).map { "w0$it" }, plan.newCards)
    }

    // v1 calibration restored: day one introduces 8 cards under the default config
    // (synonyms rotate presentations, they never multiply schedules).
    @Test
    fun dayOneIntroducesEightCards() {
        var state = Box.state((1..20).map { Box.word(it, synonyms = listOf("s$it")) })
        val plan = Box.candidates(state)
        assertEquals((1..8).map { "w0$it" }, plan.newCards)

        for (id in plan.newCards) {
            state = Box.answered(state, id, Rating.Good, now)
        }
        assertEquals(8, state.newIntroduced["2026-07-01"])
        assertEquals(8, BoxEngine.statistics(state, now, Box.TZ).activeCount)
        val ninth = BoxEngine.answer(state, "w09", Rating.Good, now, Box.TZ)
        assertEquals(AnswerStatus.DroppedPoolFull, ninth.status)
    }

    @Test
    fun relearningShareGateBlocksAtTwentyPercentOfActiveCards() {
        var state = Box.state((1..12).map { Box.word(it) })
        val future = Box.plusDays(now, 5.0)
        val past = Box.plusDays(now, -1.0)
        for (n in 1..8) {
            state = Box.inject(state, Box.sched("w0$n", dueMillis = future, lastReviewMillis = past))
        }
        for (n in 9..10) {
            state = Box.inject(
                state,
                Box.sched("w" + n.toString().padStart(2, '0'), phase = CardPhase.Relearning, dueMillis = future, lastReviewMillis = past),
            )
        }
        // 2 of 10 active cards relearning = 20% → gate closed
        val blocked = Box.candidates(state)
        assertTrue(blocked.newCards.isEmpty())
        assertTrue(blocked.unlockedPhrases.isEmpty())
        assertEquals(0, BoxEngine.statistics(state, now, Box.TZ).newSlotsAvailable)

        // Drop to 2 of 11 (18%) → gate open; seed-order growth resumes.
        val healed = Box.inject(state, Box.sched("w11", dueMillis = future, lastReviewMillis = past))
        assertEquals(listOf("w12"), Box.candidates(healed).newCards)
    }

    @Test
    fun relearningSubGatePassesBelowTenActiveCards() {
        var state = Box.state((1..8).map { Box.word(it) })
        val future = Box.plusDays(now, 5.0)
        val past = Box.plusDays(now, -1.0)
        for (n in 1..2) {
            state = Box.inject(
                state,
                Box.sched("w0$n", phase = CardPhase.Relearning, dueMillis = future, lastReviewMillis = past),
            )
        }
        for (n in 3..5) {
            state = Box.inject(state, Box.sched("w0$n", dueMillis = future, lastReviewMillis = past))
        }
        // 2 of 5 relearning = 40%, but < 10 active cards → still introduces.
        assertEquals((6..8).map { "w0$it" }, Box.candidates(state).newCards)
    }

    @Test
    fun backlogGateBlocksOnProjectedPostSessionBacklog() {
        var state = Box.state((1..70).map { Box.word(it) })
        for (n in 1..61) {
            val id = "w" + n.toString().padStart(2, '0')
            state = Box.inject(
                state,
                Box.sched(id, dueMillis = now - n * 60_000L, lastReviewMillis = Box.plusDays(now, -1.0)),
            )
        }
        // 61 due cards − 30 sessionCap = 31 ≥ dueSoftCap 30 → closed.
        assertEquals(61, BoxEngine.dueNow(state, now).size)
        assertTrue(Box.candidates(state).newCards.isEmpty())
    }

    @Test
    fun budgetTracksLearningLoadAcrossRecomposition() {
        var state = Box.state((1..10).map { Box.word(it) }, Box.config(maxLearning = 5))
        val plan = Box.candidates(state)
        assertEquals(5, plan.newCards.size)

        for (id in plan.newCards.take(3)) {
            state = Box.answered(state, id, Rating.Good, now)
        }
        assertEquals(listOf("w04", "w05"), Box.candidates(state).newCards)

        for (id in Box.candidates(state).newCards) {
            state = Box.answered(state, id, Rating.Good, now)
        }
        assertTrue(Box.candidates(state).newCards.isEmpty()) // pool full: 5 cards in learning
    }

    @Test
    fun poolRefillsOnGraduation() {
        var state = Box.state((1..10).map { Box.word(it) }, Box.config(maxLearning = 3))
        for (n in 1..3) {
            state = Box.answered(state, "w0$n", Rating.Good, now)
        }
        assertTrue(Box.candidates(state).newCards.isEmpty())

        // Graduating w01 frees a slot; the next seed-order card fills it.
        val step = Box.plusSeconds(now, 700)
        state = Box.answered(state, "w01", Rating.Good, step)
        assertEquals(listOf("w04"), Box.candidates(state, nowMillis = step).newCards)
    }

    @Test
    fun candidateSelectionIsPureAndDeterministic() {
        val state = Box.state((1..10).map { Box.word(it) })
        val first = Box.candidates(state)
        repeat(5) { assertEquals(first, Box.candidates(state)) }
        assertTrue(state.scheduling.isEmpty())
        assertTrue(state.newIntroduced.isEmpty())
    }

    @Test
    fun enqueuedLeadWithinBudgetAndPhrasePullsComponentsFirst() {
        var state = Box.state((1..10).map { Box.word(it) }, Box.config(maxLearning = 5))
        state = BoxEngine.enqueue(state, listOf("w07"))
        assertEquals(
            listOf("w07", "w01", "w02", "w03", "w04"),
            Box.candidates(state).newCards,
        )

        var withPhrase = Box.state(
            (1..6).map { Box.word(it) } + Box.phrase("p1", components = listOf("w05", "w06")),
            Box.config(maxLearning = 5),
        )
        withPhrase = BoxEngine.enqueue(withPhrase, listOf("p1"))
        assertEquals(listOf("w05", "w06", "p1"), withPhrase.enqueued)
        // Locked phrase never enters, even enqueued; components lead, then automatic
        // growth fills the rest of the pool budget.
        assertEquals(
            listOf("w05", "w06", "w01", "w02", "w03"),
            Box.candidates(withPhrase).newCards,
        )
    }

    @Test
    fun enqueuedRespectLoadThrottlePackDripsIn() {
        var state = Box.state((1..10).map { Box.word(it) }, Box.config(maxLearning = 2))
        state = BoxEngine.enqueue(state, listOf("w06", "w07", "w08"))
        assertEquals(listOf("w06", "w07"), Box.candidates(state).newCards)

        state = Box.answered(state, "w06", Rating.Good, now)
        state = Box.answered(state, "w07", Rating.Good, now)
        assertTrue(Box.candidates(state).newCards.isEmpty())
        assertEquals(listOf("w08"), state.enqueued) // still waiting for a slot
    }

    @Test
    fun enqueueSkipsUnknownScheduledAndDuplicates() {
        var state = Box.state(listOf(Box.word(1), Box.word(2)))
        state = Box.answered(state, "w01", Rating.Good, now)
        state = BoxEngine.enqueue(state, listOf("w01", "zzz", "w02", "w02"))
        assertEquals(listOf("w02"), state.enqueued)
    }
}
